crates/modules/bank/src/lib.rs

```diff
@@ -0,0 +1,592 @@
+            denom: &str,
+            amount: u128,
+        ) -> Result<()>;
+    }
+
+    /// An event emitted when a coin is sent to another account.
+    #[derive(SchemaValue, Clone, Default)]
+    #[non_exhaustive]
+    pub struct EventSend<'a> {
+        /// The account that sent the coin.
+        pub from: AccountID,
+        /// The account that received the coin.
+        pub to: AccountID,
+        /// The coin that was sent.
+        pub coin: Coin<'a>,
+    }
+
+    /// An event emitted when a coin is minted for an account.
+    #[derive(SchemaValue, Clone, Default)]
+    #[non_exhaustive]
+    pub struct EventMint<'a> {
+        /// The account that minted the coin.
+        pub to: AccountID,
+        /// The coin that was minted.
+        pub coin: Coin<'a>,
+    }
+
+    /// An event emitted when a coin is burned from an account.
+    #[derive(SchemaValue, Default, Clone)]
+    #[non_exhaustive]
+    pub struct EventBurn<'a> {
+        /// The account that burned the coin.
+        pub from: AccountID,
+        /// The coin that was burned.
+        pub coin: Coin<'a>,
+    }
+
+    /// The bank module used for transfering tokens between accounts.
+    impl Bank {
+        /// Called when the bank module is created.
+        #[on_create]
+        pub fn create(&self, ctx: &mut Context) -> Result<()> {
+            self.super_admin.set(ctx, ctx.caller())?;
+            Ok(())
+        }
+
+        /// Create a new denom.
+        #[publish]
+        pub fn create_denom(&self, ctx: &mut Context, denom: &str, admin: AccountID) -> Result<()> {
+            ensure!(self.super_admin.get(ctx)? == ctx.caller(), "not authorized");
+            self.denom_admins.set(ctx, denom, admin)?;
+            Ok(())
+        }
+
+        /// Set the global send hook.
+        #[publish]
+        pub fn set_global_send_hook(&self, ctx: &mut Context, hook: AccountID) -> Result<()> {
+            ensure!(self.super_admin.get(ctx)? == ctx.caller(), "not authorized");
+            self.global_send_hook.set(ctx, hook)?;
+            Ok(())
+        }
+
+        /// Set the denom send hook.
+        #[publish]
+        pub fn set_denom_send_hook(
+            &self,
+            ctx: &mut Context,
+            denom: &str,
+            hook: AccountID,
+        ) -> Result<()> {
+            // Only denom admin can set send hooks
+            let admin = self
+                .denom_admins
+                .get(ctx, denom)?
+                .ok_or(error!("denom not defined"))?;
+            ensure!(admin == ctx.caller(), "not authorized");
+            self.denom_send_hooks.set(ctx, denom, hook)?;
+            Ok(())
+        }
+
+        /// Set the denom recieve hook.
+        #[publish]
+        pub fn set_denom_recieve_hook(&self, ctx: &mut Context, hook: AccountID) -> Result<()> {
+            // Only denom admin can set recieve hooks
+            let caller = ctx.caller();
+            self.denom_recieve_hooks.set(ctx, caller, hook)?;
```

Comment:
So we don't call receive hook by default? Only when it's configured?

If the hook account is is empty we should delete the hook